src/network/networkd-dhcp4.c

```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
         return dhcp4_request_address_and_routes(link, true);
 }
 
+static int dhcp4_validate_persistent_lease(Link *link, sd_dhcp_lease *lease) {
+        assert(link);
+        assert(link->manager);
+        assert(lease);
+
+        usec_t lifetime, timestamp_realtime, expiration_realtime;
+        usec_t now_realtime;
+        usec_t remaining_lifetime;
+        int r;
+
+        r = sd_dhcp_lease_get_lifetime(lease, &lifetime);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get lease lifetime: %m");
+
+
+        r = sd_dhcp_lease_get_timestamp(lease, CLOCK_REALTIME, &timestamp_realtime);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get realtime timestamp: %m");
+
+        expiration_realtime = timestamp_realtime + lifetime;
+
+        now_realtime = now(CLOCK_REALTIME);
+
+        /*checking if lease is expired */
+        if (expiration_realtime <= now_realtime) {
+                if (FLAGS_SET(link->network->dhcp_client_persist_leases, DHCP_CLIENT_PERSIST_LEASES_EXPIRED)) {
+                        log_link_info(link, "Saved DHCP lease expired, but will be used.");
+                        //Extending lease by 1 minute
+                        //TODO: Might need to set some fuzzing/randomness to this value. Congestion could
+                        //occur at scale
+                        remaining_lifetime = 1 * USEC_PER_MINUTE;
+                }
+                else {
+                        log_link_info(link, "Saved DHCP lease expired, not using.");
+                        return 0;
+                }
+        }
+        else {
+                remaining_lifetime = lifetime - (now_realtime - timestamp_realtime);
+        }
+
+        r = sd_dhcp_client_update_lease_lifetime(link->dhcp_client, lease, remaining_lifetime);
+
+        if (r < 0)
+                log_link_warning_errno(link, r, "Failed to update the lease lifetime: %m");
+
+        return 1;
+}
+
+static int dhcp4_load_persistent_lease(Link *link) {
+        _cleanup_(sd_dhcp_lease_unrefp) sd_dhcp_lease *lease = NULL;
+        _cleanup_free_ char *lease_file = NULL;
+        _cleanup_free_ char * lease_path = NULL;
+        int r;
+        int dir_fd;
+
+        assert(link);
+        assert(link->network);
+
+        r = link_get_dhcp_client_lease_path(link, &dir_fd, &lease_file);
+
+        /* persistent storage not avaiable/not configured, ignore */
+        if (r == -EBUSY || r == 0)
+                return r;
+
+        if (r < 0) {
+                log_link_debug_errno(link, r, "Failed to get persistent lease path: %m");
+                return r;
+        }
+
+        if (dir_fd >= 0) {
+                lease_path = path_join("/var/lib/systemd/network", lease_file);
```

Comment:
Please suffix with `/` for the directory, even `path_join()` gracefully handle that.